kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/data/ColumnVectorConverter.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal.data;
+
+import java.math.BigDecimal;
+import java.util.Optional;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.types.*;
+import io.delta.kernel.defaults.internal.data.vector.*;
+
+public class ColumnVectorConverter {
+    public static ColumnVector convertToOutputType(ColumnVector c, DataType outputType) {
+        DataType inputType = c.getDataType();
+        if (inputType == outputType) {
+            return c;
+        }
+        if (outputType instanceof BooleanType) {
+            // TODO: how to convert string to boolean?
+            // "true" "TRUE", "T" -> true ?
+            throw new IllegalStateException("convert to boolean type is not yet supported");
+        } else if (outputType instanceof ShortType) {
+            return convertToShort(c);
+        }  else if (outputType instanceof IntegerType) {
+            return convertToInteger(c);
+        } else if (outputType instanceof LongType) {
+            return convertToLong(c);
+        } else if (outputType instanceof DecimalType) {
+            return convertToDecimal(c);
+        } else if (outputType instanceof FloatType) {
+            return convertToFloat(c);
+        } else if (outputType instanceof DoubleType) {
+            return convertToDouble(c);
+        } else {
```

Comment:
Its not clear to me if complex type can be recursive from here https://spark.apache.org/docs/latest/sql-ref-datatypes.html , I am guessing yes, but would like to confirm with code reviewer thats the case before handling complex least common type resolution for its element

> *** For a complex type, the precedence rule applies recursively to its component elements.